src/gui/guiFormSpecMenu.cpp

```diff
@@ -4124,6 +4124,13 @@ bool GUIFormSpecMenu::preprocessEvent(const SEvent& event)
 		return handled;
 	}
 
+#ifdef HAVE_TOUCHSCREENGUI
+	if (m_second_touch) {
+		OnEvent(event);
+		return true;
```

Comment:
What does this do?

I applied this patch to understand how a "second touch" event is processed:
```patch
diff --git a/src/client/inputhandler.cpp b/src/client/inputhandler.cpp
index 7b54ac93d6..27f8093925 100644
--- a/src/client/inputhandler.cpp
+++ b/src/client/inputhandler.cpp
@@ -98,6 +98,10 @@ void KeyCache::populate()
 
 bool MyEventReceiver::OnEvent(const SEvent &event)
 {
+	if (event.EventType == irr::EET_TOUCH_INPUT_EVENT &&
+			event.TouchInput.Event == irr::ETIE_LEFT_UP && event.TouchInput.ID == 1) {
+		errorstream << "[MyEventReceiver::OnEvent] got second touch" << std::endl;
+	}
 	/*
 		React to nothing here if a menu is active
 	*/
diff --git a/src/gui/guiFormSpecMenu.cpp b/src/gui/guiFormSpecMenu.cpp
index 5f6f763714..b3f6d57bd2 100644
--- a/src/gui/guiFormSpecMenu.cpp
+++ b/src/gui/guiFormSpecMenu.cpp
@@ -4047,6 +4047,10 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode)
 
 bool GUIFormSpecMenu::preprocessEvent(const SEvent& event)
 {
+	if (event.EventType == irr::EET_TOUCH_INPUT_EVENT &&
+			event.TouchInput.Event == irr::ETIE_LEFT_UP && event.TouchInput.ID == 1) {
+		errorstream << "[GUIFormSpecMenu::preprocessEvent] got second touch" << std::endl;
+	}
 	// The IGUITabControl renders visually using the skin's selected
 	// font, which we override for the duration of form drawing,
 	// but computes tab hotspots based on how it would have rendered
@@ -4126,6 +4130,7 @@ bool GUIFormSpecMenu::preprocessEvent(const SEvent& event)
 
 #ifdef HAVE_TOUCHSCREENGUI
 	if (m_second_touch) {
+		std::cout << "[GUIFormSpecMenu::preprocessEvent] DOING SPECIAL HANDLING OF SECOND TOUCH" << std::endl;
 		OnEvent(event);
 		return true;
 	}
@@ -4147,6 +4152,10 @@ void GUIFormSpecMenu::tryClose()
 
 bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 {
+	if (event.EventType == irr::EET_TOUCH_INPUT_EVENT &&
+			event.TouchInput.Event == irr::ETIE_LEFT_UP && event.TouchInput.ID == 1) {
+		errorstream << "[GUIFormSpecMenu::OnEvent] got second touch" << std::endl;
+	}
 	if (event.EventType==EET_KEY_INPUT_EVENT) {
 		KeyPress kp(event.KeyInput);
 		if (event.KeyInput.PressedDown && (
@@ -4314,8 +4323,10 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 		// The second touch (see GUIModalMenu::preprocessEvent() function)
 		ButtonEventType touch = BET_OTHER;
 		if (event.EventType == EET_TOUCH_INPUT_EVENT) {
-			if (event.TouchInput.Event == ETIE_LEFT_UP)
+			if (event.TouchInput.Event == ETIE_LEFT_UP) {
+				errorstream << "[GUIFormspecMenu::OnEvent] handling second touch" << std::endl;
 				touch = BET_RIGHT;
+			}
 		}
 #endif
 
diff --git a/src/gui/modalMenu.cpp b/src/gui/modalMenu.cpp
index 0f5ec787a3..6a90f1a0c9 100644
--- a/src/gui/modalMenu.cpp
+++ b/src/gui/modalMenu.cpp
@@ -235,6 +235,11 @@ void GUIModalMenu::leave()
 
 bool GUIModalMenu::preprocessEvent(const SEvent &event)
 {
+	if (event.EventType == irr::EET_TOUCH_INPUT_EVENT &&
+			event.TouchInput.Event == irr::ETIE_LEFT_UP && event.TouchInput.ID == 1) {
+		errorstream << "[GUIModalMenu::preprocessEvent] got second touch" << std::endl;
+	}
+
 #ifdef __ANDROID__
 	// clang-format off
 	// display software keyboard when clicking edit boxes
@@ -301,6 +306,7 @@ bool GUIModalMenu::preprocessEvent(const SEvent &event)
 			auto focused = Environment->getFocus();
 			if (!focused)
 				return true;
+			errorstream << "[GUIModalMenu::preprocessEvent] detected second touch, propagating" << std::endl;
 			// The second-touch event is propagated as is (not converted).
 			m_second_touch = true;
 			focused->OnEvent(event);
```

Each successful "second tap" resulted in this output:

```
[MyEventReceiver::OnEvent] got second touch
[GUIFormSpecMenu::preprocessEvent] got second touch
[GUIModalMenu::preprocessEvent] got second touch
[GUIModalMenu::preprocessEvent] detected second touch, propagating
[GUIFormSpecMenu::OnEvent] got second touch
[GUIFormspecMenu::OnEvent] handling second touch
```

So, this part of the code doesn't seem to have any effect. Can it be removed? Or is it needed to handle a special case?

Reply:
Yes, that part can be removed. Thank you for noticing it.

It was written when I was trying to understand the relation between `GUIFormSpecMenu` class and `GUIModalMenu` class, but it should not have any function now.